File: 80.py
# task_scheduler.py
import heapq
from collections import Counter, deque

def least_interval(tasks, n):
    if n == 0:
        return len(tasks)
    freq = Counter(tasks)
    # max-heap using negative counts
    heap = [-c for c in freq.values()]
    heapq.heapify(heap)
    time = 0
    while heap:
        temp = []
        k = n + 1
        for _ in range(k):
            if heap:
                c = heapq.heappop(heap)
                if c + 1 < 0:  # still tasks remaining
                    temp.append(c+1)
                time += 1
            else:
                # if nothing to do but there are pending tasks in temp, we count idle
                if temp:
                    time += 1
                else:
                    break
        for item in temp:
            heapq.heappush(heap, item)
    return time

# Driver
if __name__ == "__main__":
    tests = [
        (["A","A","A","B","B","B"], 2, 8),
        (["A","A","A","B","B","B"], 0, 6),
        (["A","A","A","A","A","A","B","C","D","E","F","G"], 2, 16)
    ]
    for tasks, n, expected in tests:
        out = least_interval(tasks, n)
        print("tasks:", tasks, "n:", n, "->", out, " expected:", expected)
        assert out == expected
    print("Python Task Scheduler tests passed.")

File: 80.js
// task_scheduler.js
class MaxHeap {
  constructor() {
    this.data = [];
  }
  size(){ return this.data.length; }
  push(val){
    this.data.push(val);
    this._siftUp(this.data.length - 1);
  }
  pop(){
    if (this.data.length === 0) return undefined;
    const top = this.data[0];
    const last = this.data.pop();
    if (this.data.length) {
      this.data[0] = last;
      this._siftDown(0);
    }
    return top;
  }
  _siftUp(i){
    while (i > 0) {
      const p = Math.floor((i-1)/2);
      if (this.data[i] > this.data[p]) {
        [this.data[i], this.data[p]] = [this.data[p], this.data[i]];
        i = p;
      } else break;
    }
  }
  _siftDown(i){
    const n = this.data.length;
    while (true) {
      let largest = i;
      const l = 2*i + 1, r = 2*i + 2;
      if (l < n && this.data[l] > this.data[largest]) largest = l;
      if (r < n && this.data[r] > this.data[largest]) largest = r;
      if (largest !== i) {
        [this.data[i], this.data[largest]] = [this.data[largest], this.data[i]];
        i = largest;
      } else break;
    }
  }
}

function leastInterval(tasks, n) {
  if (n === 0) return tasks.length;
  const freq = new Map();
  for (const t of tasks) freq.set(t, (freq.get(t) || 0) + 1);
  const heap = new MaxHeap();
  for (const c of freq.values()) heap.push(c);

  let time = 0;
  while (heap.size()) {
    const temp = [];
    let k = n + 1;
    for (let i = 0; i < k; ++i) {
      if (heap.size()) {
        let top = heap.pop();
        top -= 1;
        if (top > 0) temp.push(top);
        time += 1;
      } else {
        if (temp.length) time += 1;
        else break;
      }
    }
    for (const x of temp) heap.push(x);
  }
  return time;
}

// Driver
const tests = [
  {tasks: ["A","A","A","B","B","B"], n:2, expect:8},
  {tasks: ["A","A","A","B","B","B"], n:0, expect:6},
  {tasks: ["A","A","A","A","A","A","B","C","D","E","F","G"], n:2, expect:16}
];

for (const t of tests) {
  const out = leastInterval(t.tasks, t.n);
  console.log("tasks:", t.tasks.join(''), "n:", t.n, "->", out, "expected:", t.expect);
}
console.log("JS Task Scheduler tests done.");
